src/function/table/call/hnsw/query_hnsw_index.cpp

```diff
@@ -0,0 +1,140 @@
+                                 ->getTableCatalogEntry(context->getTransaction(),
+                                     this->boundInput.indexEntry->getUpperRelTableID())
+                                 ->ptrCast<catalog::RelTableCatalogEntry>();
+    lowerHNSWRelTableEntry = context->getCatalog()
+                                 ->getTableCatalogEntry(context->getTransaction(),
+                                     this->boundInput.indexEntry->getLowerRelTableID())
+                                 ->ptrCast<catalog::RelTableCatalogEntry>();
+}
+
+static std::unique_ptr<TableFuncBindData> bindFunc(main::ClientContext* context,
+    const TableFuncBindInput* input) {
+    const auto indexName = input->getLiteralVal<std::string>(0);
+    const auto tableName = input->getLiteralVal<std::string>(1);
+    const auto& queryVal = input->params[2]->constCast<binder::LiteralExpression>().getValue();
+    const auto k = input->getLiteralVal<int64_t>(3);
+
+    auto& nodeTableEntry = storage::IndexUtils::bindTable(*context, tableName, indexName,
+        storage::IndexOperation::QUERY);
+    const auto indexEntry = common::ku_dynamic_cast<catalog::HNSWIndexCatalogEntry*>(
+        context->getCatalog()->getIndex(context->getTransaction(), nodeTableEntry.getTableID(),
+            indexName));
+    auto& indexColumnType = nodeTableEntry.getProperty(indexEntry->getIndexColumnName()).getType();
+    KU_ASSERT(indexColumnType.getLogicalTypeID() == common::LogicalTypeID::ARRAY);
+    const auto dimension =
+        indexColumnType.getExtraTypeInfo()->constPtrCast<common::ArrayTypeInfo>()->getNumElements();
+    storage::HNSWIndexUtils::validateQueryVector(queryVal.getDataType(), dimension);
+
+    std::vector<float> queryVector;
+    KU_ASSERT(queryVal.getChildrenSize() > 0);
+    queryVector.resize(queryVal.getChildrenSize());
+    for (auto i = 0u; i < queryVector.size(); i++) {
+        queryVector[i] = common::NestedVal::getChildVal(&queryVal, i)->getValue<float>();
+    }
+
+    auto outputNode = input->binder->createQueryNode("nn", {&nodeTableEntry});
+    input->binder->addToScope(outputNode->toString(), outputNode);
+    binder::expression_vector columns;
+    columns.push_back(outputNode->getInternalID());
+    columns.push_back(input->binder->createVariable("_distance", common::LogicalType::DOUBLE()));
+    auto boundInput = BoundQueryHNSWIndexInput{&nodeTableEntry, indexEntry, std::move(queryVector),
+        static_cast<uint64_t>(k)};
+    auto config = storage::QueryHNSWConfig{input->optionalParams};
+    return std::make_unique<QueryHNSWIndexBindData>(context, std::move(columns), boundInput, config,
+        outputNode);
+}
+
+static common::offset_t tableFunc(const TableFuncInput& input, TableFuncOutput& output) {
+    const auto localState = input.localState->ptrCast<QueryHNSWLocalState>();
+    const auto bindData = input.bindData->constPtrCast<QueryHNSWIndexBindData>();
+    if (!localState->result.has_value()) {
```

Comment:
Maybe comment that you have the assumption that the localState->result cannot exist if we have not yet searched. You could also be more explicit and add a boolean variable, e.g., localState->searchPerformed.